### ISSUE TITLE
Remove possible truncation at implicit conversion to type "unsigned char"

### ISSUE THREAD

**Issue by vflouriot** [kokke/tiny-AES-c#118]
Inside aes.c line 189 : possible truncation at implicit conversion to type "unsigned char"

// Function RotWord()
      {
        k = tempa[0];
        tempa[0] = tempa[1];
        tempa[1] = tempa[2];
        tempa[2] = tempa[3];
        **tempa[3] = k;**
      }

**Comment by kokke** [comment]
I don't understand the warning you're quoting - tempa[3] is already an unsigned char to begin with. Sounds spurious... Line 155:  ` uint8_t tempa[4]; // Used for the column/row operations`

What tool gave you this warning?

**Comment by vflouriot** [comment]
Hello Kokke,

The compiler is Tasking Tricore v6.2r2

"k" is defined without type, and as per C language "int" is selected as default, witch is a non portable type (int is 32 bit wide on Aurix).

So the compiler is right to warn because allocating k to tempa[3] may cause a loss of information.

I do think you should have wrote : tempa[3] = (uint8_t)k;

Vincent

**Comment by Nable80** [comment]
Strictly speaking, ``k`` was defined with a type (unsigned [int]). It still makes sense to use another temporary variable here, some variable with a right type (``uint8_t``) to avoid expansion in line ``k = tempa[0]`` that may cost something for 8-bit MCUs.

**Comment by vflouriot** [comment]
Hi Kokke you are both right and wrong.

The compiler warning message does not says there is a loss of information, it says that this line could cause loss of information depending of a context he's not able to check by itself.

He is definitely right to warn. Allocate a int into a byte do may cause a loss of information.

Once you have checked the content of k will always fit into a unint_8 you need to tell the compiler that you did the check and you agree this variable should be **explicitely casted** from int to unint_8.

Writing (uint8_t)k instead of k will not change any bit into generated machine code but will just prevent the compiler issuing the warning.

Explicit casting makes somewhere a cleaner code because you show that have checked and you you agree to cast. In some other conditions this could cause a real bug.

**Comment by DamonHD** [comment]
I agree that the cast should be added so that the compiler knows that you know that it is OK to truncate to 8 bits.

(I spend too much of my life porting C/C++ code between platforms, giving clear heads up to other programmers and the compilers that you are clear what is happening is good.)

Rgds

Damon

**Comment by Nable80** [comment]
> The code assigning to k basically looks like this: [...]

This is not the place where compiler complains. Questionable assignments are here: <code>k = tempa[0]; ... tempa[3] = k;</code> Of course, data loss isn't possible here but reusing ``k`` as a byte storage in this place is questionable (both due to different size and different meaning of variables).

**Comment by kokke** [comment]
This commit hopefully solved all grief :)

https://github.com/kokke/tiny-AES-c/commit/691aa01d4ab80c3dcc50f36408f0383f785e1977#diff-9789999dc6dd261657c63f757d1e680d

**Comment by vflouriot** [comment]
Hi Kokke thanks a lot this definitely remove the warning at my side.

Vincent

**Comment by kokke** [comment]
@vflouriot I never did say thank you for reporting this issue, but thanks for taking an interest in improving the code base :)
